src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -473,29 +474,9 @@ public void OnAbort(Exception abortException)
                 }
             }
 
-            public void OnRefused()
+            public void OnRefused(Exception abortException = null)
             {
-                bool signalWaiter;
-                lock (SyncObject)
-                {
-                    if (NetEventSource.IsEnabled) Trace("");
-
-                    if (_disposed || _state == StreamState.Aborted)
-                    {
-                        return;
-                    }
-
-                    _state = StreamState.Aborted;
-                    _canRetry = true;
-
-                    signalWaiter = _hasWaiter;
-                    _hasWaiter = false;
-                }
-
-                if (signalWaiter)
-                {
-                    _waitSource.SetResult(true);
-                }
+                OnAbort(abortException, canRetry: true);
```

Comment:
I originally separated out OnAbort and OnRefused as two different methods to emphasize that they are different operations, to try to avoid one being called when the other should. Also, originally OnRefused didn't take an abortException because technically it's not aborting, and the user should never see this exception. And while there's some small amount of duplicate code, it's not much and it's basically the same boilerplate code that all the On* routines have, so it's not a big deal.

I don't really have a strong opinion here; if you want to merge these routines, go for it. But if you're going to merge them, it seems unnecessary to have an OnRefused routine at all; just call OnAbort directly.

Reply:
I'd prefer to keep similar code as one to avoid state where we fix bug in one of them but not the other. Will merge into one

Reply:
Ok